flink-table/flink-table-planner-blink/src/test/java/org/apache/flink/table/planner/catalog/CatalogStatisticsTest.java

```diff
@@ -59,32 +63,75 @@
 	).build();
 
 	@Test
-	public void testGetStatsFromCatalog() throws Exception {
+	public void testGetStatsFromCatalogForConnectorCatalogTable() throws Exception {
 		EnvironmentSettings settings = EnvironmentSettings.newInstance().useBlinkPlanner().inBatchMode().build();
 		TableEnvironment tEnv = TableEnvironment.create(settings);
 		Catalog catalog = tEnv.getCatalog(tEnv.getCurrentCatalog()).orElse(null);
 		assertNotNull(catalog);
 		catalog.createTable(
-			ObjectPath.fromString("default_database.T1"),
-			ConnectorCatalogTable.source(new TestTableSource(true, tableSchema), true),
-			false);
+				ObjectPath.fromString("default_database.T1"),
+				ConnectorCatalogTable.source(new TestTableSource(true, tableSchema), true),
+				false);
 		catalog.createTable(
-			ObjectPath.fromString("default_database.T2"),
-			ConnectorCatalogTable.source(new TestTableSource(true, tableSchema), true),
-			false);
+				ObjectPath.fromString("default_database.T2"),
+				ConnectorCatalogTable.source(new TestTableSource(true, tableSchema), true),
+				false);
 
+		alterTableStatistics(catalog);
+
+		Table table = tEnv.sqlQuery("select * from T1, T2 where T1.s3 = T2.s3");
+		String result = tEnv.explain(table);
+		// T1 is broadcast side
+		String expected = TableTestUtil.readFromResource("/explain/testGetStatsFromCatalogForConnectorCatalogTable.out");
+		assertEquals(expected, TableTestUtil.replaceStageId(result));
+	}
+
+	@Test
+	public void testGetStatsFromCatalogForCatalogTableImpl() throws Exception {
+		EnvironmentSettings settings = EnvironmentSettings.newInstance().useBlinkPlanner().inBatchMode().build();
+		TableEnvironment tEnv = TableEnvironment.create(settings);
+		Catalog catalog = tEnv.getCatalog(tEnv.getCurrentCatalog()).orElse(null);
+		assertNotNull(catalog);
+
+		Map<String, String> properties = new HashMap<>();
+		properties.put("connector.type", "filesystem");
+		properties.put("connector.property-version", "1");
+		properties.put("connector.path", "/path/to/csv");
+
+		properties.put("format.type", "csv");
+		properties.put("format.property-version", "1");
+		properties.put("format.field-delimiter", ";");
+
+		// schema
+		DescriptorProperties descriptorProperties = new DescriptorProperties(true);
+		descriptorProperties.putTableSchema("format.fields", tableSchema);
+		properties.putAll(descriptorProperties.asMap());
+
+		catalog.createTable(
+				ObjectPath.fromString("default_database.T1"),
+				new CatalogTableImpl(tableSchema, properties, ""),
+				false);
+		catalog.createTable(
+				ObjectPath.fromString("default_database.T2"),
+				new CatalogTableImpl(tableSchema, properties, ""),
+				false);
+
+		alterTableStatistics(catalog);
+
+		Table table = tEnv.sqlQuery("select * from T1, T2 where T1.s3 = T2.s3");
+		String result = tEnv.explain(table);
+		// T1 is broadcast side
```

Comment:
Do we have other choices for validating the table stats? Relying on small table will be broadcasted in a join is kind of fragile IMO.

Reply:
i think one choice is using metadata handlers to verify the table stats.

Reply:
Yes, let's test this via metadata handlers.